source/mir/math/stat.d

```diff
@@ -23,40 +23,160 @@ import mir.math.common: fmamath;
 import mir.math.sum;
 import mir.primitives;
 import std.range.primitives: isInputRange;
-import std.traits: isArray, isFloatingPoint;
+import std.traits: isArray, isFloatingPoint, isMutable;
+
+/++
+Output range for mean.
++/
+struct MeanAccumulator(T, Summation summation) 
+    if (isMutable!T)
+{
+    private:
+    
+    Summator!(T, summation) sum;
+    size_t count;
+    
+    public:
+    
+    this()(T x, size_t n = 0) {
+        sum = Summator!(T, summation)(x);
+        count = n;
+    }
+    
+    F mean(F = T)() @property
+    {
+        return sum.sum / cast(F) count;
+    }
+    
+    void put(U)(U x)
+    {
+        this.sum.put(x);
+		size_t n;
+        static if (hasShape!U) {
+        	n = x.elementCount;
+        } else static if (hasLength!U) {
+        	n = x.length;
+        } else {
+            n = 1;    
+        }
+        this.count += n;
+    }
+}
+
+///
+version(mir_test) @safe pure nothrow unittest {
+    import mir.ndslice.slice : sliced;
+
+    MeanAccumulator!(double, Summation.pairwise) x;
+    x.put([0.0, 1, 2, 3, 4].sliced);
+    assert(x.mean == 2);
+    x.put(5);
+    assert(x.mean == 2.5);
+}
+
+version(mir_test) @safe pure nothrow unittest {
+    import mir.ndslice.slice : sliced;
+
+    auto x = MeanAccumulator!(float, Summation.pairwise)(0f, 0);
+    x.put([0, 1, 2, 3, 4].sliced);
+    assert(x.mean == 2);
+    x.put(5);
+    assert(x.mean == 2.5);
+}
+
+private template MeanAlgo(Summation summation, Range, F)
+{
+    static if (summation == Summation.precise)
+        alias MeanAlgo = sumPrecise!(Range, F);
+    else
+    static if (summation == Summation.appropriate)
+    {
+        static if (isSummable!(Range, F))
+            alias MeanAlgo = MeanAlgo!(Summation.pairwise, Range, F);
+        else
+        static if (is(F == class) || is(F == struct) || is(F == interface))
+            alias MeanAlgo = MeanAlgo!(Summation.naive, Range, F);
+        else
+            alias MeanAlgo = MeanAlgo!(Summation.fast, Range, F);
+    }
```

Comment:
it is better to add a template that will fetch a proper SummationType for a triple `Summation summation, Range, F `